src/marker.rs

```diff
@@ -522,26 +337,282 @@ impl<'py> Python<'py> {
     ///
     /// fn parallel_print(py: Python<'_>) {
     ///     let s = PyString::new(py, "This object cannot be accessed without holding the GIL >_<");
-    ///     py.allow_threads(move || {
+    ///     py.safe_allow_threads(move || {
     ///         println!("{:?}", s); // This causes a compile error.
     ///     });
     /// }
     /// ```
     ///
+    /// # Example: The `send_wrapper` loophole is closed by running the closure on dedicated thread
+    ///
+    /// ```should_panic
+    /// # use pyo3::prelude::*;
+    /// # use pyo3::types::PyString;
+    /// use send_wrapper::SendWrapper;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let string = PyString::new(py, "foo");
+    ///
+    ///     let wrapped = SendWrapper::new(string);
+    ///
+    ///     py.safe_allow_threads(|| {
+    ///         // panicks because this is not the thread which created `wrapped`
+    ///         let sneaky: &PyString = *wrapped;
+    ///         println!("{:?}", sneaky);
+    ///     });
+    /// });
+    /// ```
+    ///
     /// [`Py`]: crate::Py
     /// [`PyString`]: crate::types::PyString
     /// [auto-traits]: https://doc.rust-lang.org/nightly/unstable-book/language-features/auto-traits.html
     /// [Parallelism]: https://pyo3.rs/main/parallelism.html
-    pub fn allow_threads<T, F>(self, f: F) -> T
+    pub fn safe_allow_threads<T, F>(self, f: F) -> T
+    where
+        F: Send + FnOnce() -> T,
+        T: Send,
+    {
+        use std::mem::{replace, transmute};
+        use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+        use std::sync::Arc;
+        use std::thread::Builder;
+        use std::time::Duration;
+
+        use parking_lot::{const_mutex, Condvar, Mutex};
```

Comment:
As much as I'd love to use std's primitives for const-construction, I think 0.21 contains enough churn that we don't need to push an MRSV bump on users too. We can tidy up later on 😄